appfl/algorithm/iceadmm.py

```diff
@@ -0,0 +1,156 @@
+import logging
+
+log = logging.getLogger(__name__)
+
+from collections import OrderedDict
+from .algorithm import BaseServer, BaseClient
+from .misc import *
+
+import torch
+from torch.optim import *
+from torch.nn import CrossEntropyLoss
+from torch.utils.data import DataLoader
+import copy
+
+
+class ICEADMMServer(BaseServer):
+    def __init__(self, weights, model, num_clients, device, **kwargs):
+        super(ICEADMMServer, self).__init__(weights, model, num_clients, device)
+
+        self.__dict__.update(kwargs)
+        self.num_clients = num_clients
+        self.weights = weights
+
+        self.primal_states = OrderedDict()
+        self.dual_states = OrderedDict()
+        for i in range(num_clients):
+            self.primal_states[i] = OrderedDict()
+            self.dual_states[i] = OrderedDict()
+
+        self.penalty = OrderedDict()
+
+    def update(self, local_states: OrderedDict):
+        
+        """ Inputs """
+        global_state = self.model.state_dict()
+        primal_recover_from_local_states(self, local_states)
+        dual_recover_from_local_states(self, local_states)
+        penalty_recover_from_local_states(self, local_states)        
```

Comment:
Please consider moving these under `BaseServer`. And, these functions do pretty much the same thing.

Reply:
Reflected